[Intel MKL] Adding support for MKL builds with AVX2.

claynerobison opened the pull request:
MKL-DNN ignores the compiler switches and takes code paths based on the platform detected at runtime. This commit will add support for avx2 instructions to the rest of TensorFlow.

claynerobison committed:
Adding support for MKL builds with AVX2. MKL-DNN ignores the compiler switches and takes code paths based on the platform detected at runtime. This commit will add support for avx2 instructions to the rest of TensorFlow.

rmlarsen commented on tensorflow/tools/ci_build/linux/mkl/build-dev-container.sh:
```diff
@@ -49,5 +52,27 @@ TF_DOCKER_BUILD_TYPE="MKL" \
+  ${WORKSPACE}/tensorflow/tools/docker/parameterized_docker_build.sh
+
+# Build containers for AVX2
+TF_BAZEL_BUILD_OPTIONS="--config=mkl --copt=-mavx2 --cxxopt=-D_GLIBCXX_USE_CXX11_ABI=0"
```
Does AVX2 imply FMA and AVX? The two latter are typically more important for performance. Should we explicitly list all three?

rmlarsen commented on tensorflow/tools/ci_build/linux/mkl/build-dev-container.sh:
```diff
@@ -34,6 +34,9 @@ echo "TF_DOCKER_BUILD_DEVEL_BRANCH=${TF_DOCKER_BUILD_DEVEL_BRANCH}"
 echo "TF_DOCKER_BUILD_VERSION=${TF_DOCKER_BUILD_VERSION}"
 
+# Build containers for AVX
+#"TF_BAZEL_BUILD_OPTIONS": "'{}'" (default build option= avx)
```
Cruft?

claynerobison committed:
Adding better support for avx- and avx2-class instruction sets.

-march=sandybridge enables MMX, SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2, POPCNT, AVX, AES and PCLMUL instruction set support. See https://gcc.gnu.org/onlinedocs/gcc-5.4.0/gcc/x86-Options.html#x86-Options

-march=haswell enables MOVBE, MMX, SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2, POPCNT, AVX, AVX2, AES, PCLMUL, FSGSBASE, RDRND, FMA, BMI, BMI2 and F16C instruction set support. See https://gcc.gnu.org/onlinedocs/gcc-5.4.0/gcc/x86-Options.html#x86-Options

claynerobison commented:
@rmlarsen Clarified comments and added more complete support for the haswell and sandybridge instruction sets. The mtune parameters won't add instructions, but will tune cache sizes etc for the specified architecture.

rmlarsen commented:
@claynerobison Thanks!

rmlarsen commented:
@gunan @jhseu does this look good to you?

claynerobison committed:
Passing full sandybridge build params to the container build script

claynerobison commented:
Sorry, @jhseu @rmlarsen @gunan further internal testing revealed that we didn't have the full haswell arch build flag being passed in to the first two containers. Fixed it with the last commit.
